astropy/coordinates/sky_coordinate.py

```diff
@@ -1035,7 +1035,7 @@ def separation(self, other):
             not give the same answer in this case.
 
         For more on how to use this (and related) functionality, see the
-        examples in :doc:`/coordinates/matchsep`.
+        examples in :doc:`astropy:/coordinates/matchsep`.
```

Comment:
I haven't been following the various conversation on Sphinx xref and just by looking at this diff, I don't understand why this is needed in the core repo. Why not use the built-in Sphinx `:ref:` again? What is an example use case where this would benefit affiliated package? Thanks!

Reply:
> Why not use the built-in Sphinx :ref: again?

The link is to a `:doc:`, not a specific section, so `:ref:` can't be used. Perhaps the link could be changed entirely, but for page references `:doc:` is the Sphinx built-in.

> What is an example use case where this would benefit affiliated package?

In this case, for subclassing SkyCoord. Both `:doc: `and `:ref: `are not inter-sphinxed automatically, so any link in the SkyCoord subclass can break when building the docs (if warnings -> errors). This PR enables astropy to set up internal links *as if* they were intersphinx links, but have them be interpreted as internal links. Affiliate packages would see these links as intersphinx links. This is quite similar to `astropy-dev`, where Astropy sees it as internal (if it's the dev build), while affiliate packages see it as an intersphinx.

Reply:
I cannot find any documentation of `:doc:` in the official Sphinx doc (which is kinda ironic), can you please link me to its doc? Thanks!

As for `:ref:` intersphinx, I haven't seen a problem downstream where I use `~astropy.coordinates.SkyCoord` and so on. I did see a problem when the doc here refers to simply `SkyCoord` but the fix is to mention the full namespace in core API doc.

As for referring to page vs section, I don't see the point of the former. Why not have a `:ref:` at the top of the page and then use `:ref:`?

Sorry if all these had been discussed before. If so, please point me to the old discussions. Thanks for your patience!

Reply:
Actually, I do remember discussion quite a while ago that docstrings should use proper http links so that people who see them in an interactive session can actually follow them....

But in that respect the scheme with `:doc:` is better than a `:ref:` - with that, the user truly has no hope to find this except by going to the docstring online. Also, the `:doc:astrop` seems a nice clear marker. Plus, this scheme has the advantage over an `http` link that we'll notice breakage sooner.

Reply:
One question: is the `/` in front of `/coordinates` necessary? Or does it try to find the link in the local environment of the `SkyCoord` docstring rendering otherwise?

Reply:
@pllim 

> I cannot find any documentation of :doc: in the official Sphinx doc (which is kinda ironic), can you please link me to its doc? Thanks!

Took me a hot sec as well. https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html?highlight=%3Adoc%3A#cross-referencing-documents

> As for :ref: intersphinx, I haven't seen a problem downstream where I use ~astropy.coordinates.SkyCoord and so on. I did see a problem when the doc here refers to simply SkyCoord but the fix is to mention the full namespace in core API doc.

That shouldn't raise an issue. The problem, which I've encountered, is that ``:ref:`` or ``:doc:`` refer to project being built, unless the link is an intersphinx. Since there is a ``:doc:`` link in SkyCoord, when SkyCoord is subclassed the methods' docstrings can move packages due to docstring inheritance. Now the ``:ref:`` and ``:doc:`` links no longer point to Astropy and so error. Intersphinx solves this problem... but it's a bad idea to intersphinx to yourself as that doesn't get the currently built version. So this PR strips intersphinx links to Astropy when being built by Astropy, but not in affiliate packages.

> As for referring to page vs section, I don't see the point of the former. Why not have a :ref: at the top of the page and then use :ref:?

I think this is a valid point. ``:doc:`` is just a convenient ``:ref:``. I think it is actually just a ref that Sphinx handles without have a :ref: at the top.
However the above description of how a ``:doc:`` can move packages by docstring inheritance equally applies to ``:ref:``, so adding intersphinx magic is still necessary.


@mhvk 
> is the / in front of /coordinates necessary? 
I think so(?) In the docs it says "For example, if the reference :doc:`parrot` occurs in the document sketches/index, then the link refers to sketches/parrot. If the reference is :doc:`/people` or :doc:`../people`, the link refers to people." So we want to start at the base of the astropy docs.

Reply:
Maybe I should expand on the changelog entry to talk about resolving intersphinx and how this is useful for affiliate packages?

Reply:
Thanks for the explanation. It makes sense now.  The whole sphinx stuff remains a bit of black magic to me... Not sure that expanding the changelog will be needed; if anything, we'd need some docs for affiliate package maintainers, but not clear how they would find that... So, let's just stick with fixing a possible bug with subclassing `SkyCoord` here.

Reply:
Allowing a project to refer to itself via intersphinx seems generically useful for protecting from docstring inheritance. I wish sphinx did this automatically.

Reply:
https://github.com/sphinx-doc/sphinx/issues/9169

Reply:
Ah, I think I get it now. I was thinking about `:ref:` in docstring pointing to another API, but here you are talking about human doc.

Is this PR standalone or it depends on the other PR you opened over at `sphinx-astropy`?

Reply:
This PR is standalone in that it works and does the docs stuff independently.

It's definitely part of a larger push to make intersphinx and particularly xref useful for downstream packages. In some respects, this PR is the foundation on which future PRs can be built since all the downstream packages need astropy xref links to be intersphinxed, but without this PR, those links wouldn't work inside of Astropy. Now sphinx-astropy will work equally well for both Astropy and downstream packages.